File: backend/location_consent.py
"""Personel konum paylaşımı + KVKK (K / KK) onayı."""

from __future__ import annotations

from typing import Any, Optional

CONSENT_WARNING = "Bu sözleşmeleri işaretlediğinizde personel paneli (Mesaim) kullanıma açılır."

KVKK_TITLE = "KVKK aydınlatması (K)"
KVKK_TEXT = (
    "İşveren, 6698 sayılı KVKK kapsamında mesai takibi için cihazınızın konum verisini "
    "(GPS) işler. Veri giriş-çıkış ve görev/iş yeri kontrolü için kullanılır; yasal süre boyunca saklanır."
)

SHARE_TITLE = "Konum paylaşımı sözleşmesi (KK)"
SHARE_TEXT = (
    "Mesai süresince konumumun alınmasına ve konum alınamadığında yöneticimin haberdar edilmesine izin veriyorum."
)


def normalize_location_consent(raw: Any = None) -> dict:
    row = raw if isinstance(raw, dict) else {}
    kvkk = bool(row.get("accept_kvkk") or row.get("kvkk"))
    share = bool(row.get("accept_share") or row.get("kk") or row.get("location"))
    accepted = kvkk and share
    at = str(row.get("accepted_at") or "").strip() or None
    return {
        "accept_kvkk": kvkk,
        "accept_share": share,
        "accepted": accepted,
        "accepted_at": at if accepted else None,
        "required": True,
        "warning": CONSENT_WARNING,
        "kvkk_title": KVKK_TITLE,
        "kvkk_text": KVKK_TEXT,
        "share_title": SHARE_TITLE,
        "share_text": SHARE_TEXT,
    }


def location_consent_accepted(emp: Optional[dict] = None) -> bool:
    return bool(normalize_location_consent((emp or {}).get("location_consent")).get("accepted"))


def validate_location_consent(req: Optional[dict] = None) -> Optional[str]:
    row = req if isinstance(req, dict) else {}
    kvkk = bool(row.get("accept_kvkk") or row.get("kvkk"))
    share = bool(row.get("accept_share") or row.get("kk"))
    if not kvkk or not share:
        return "KVKK (K) ve konum paylaşımı (KK) sözleşmelerini işaretleyin."
    return None


def location_consent_store(req: dict, accepted_at: str) -> dict:
    return {
        "accept_kvkk": True,
        "accept_share": True,
        "accepted_at": accepted_at,
    }


def location_signal_view(emp: Optional[dict] = None) -> dict:
    row = emp or {}
    ok = row.get("location_last_ok")
    if ok is True:
        label, tone = "Konum alındı", "green"
        flag = True
    elif ok is False:
        label, tone = "Konum alınamadı", "red"
        flag = False
    else:
        label, tone = "Konum bekleniyor", "amber"
        flag = None
    return {"ok": flag, "at": row.get("location_last_at"), "label": label, "tone": tone}


def location_consent_denied_detail() -> str:
    return "KVKK ve konum paylaşımı sözleşmesini kabul etmeden Mesaim kullanılamaz."

File: frontend/src/utils/locationConsent.js
export const CONSENT_WARNING = "Bu sözleşmeleri işaretlediğinizde personel paneli (Mesaim) kullanıma açılır.";
export const CONSENT_DENIED = "KVKK ve konum paylaşımı sözleşmesini kabul etmeden Mesaim kullanılamaz.";

export const KVKK_TITLE = "KVKK aydınlatması (K)";
export const KVKK_TEXT =
  "İşveren, 6698 sayılı KVKK kapsamında mesai takibi için cihazınızın konum verisini (GPS) işler. Veri giriş-çıkış ve görev/iş yeri kontrolü için kullanılır; yasal süre boyunca saklanır.";

export const SHARE_TITLE = "Konum paylaşımı sözleşmesi (KK)";
export const SHARE_TEXT =
  "Mesai süresince konumumun alınmasına ve konum alınamadığında yöneticimin haberdar edilmesine izin veriyorum.";

export function normalizeLocationConsent(raw) {
  const row = raw && typeof raw === "object" ? raw : {};
  const accept_kvkk = Boolean(row.accept_kvkk || row.kvkk);
  const accept_share = Boolean(row.accept_share || row.kk || row.location);
  const accepted = Boolean(row.accepted ?? (accept_kvkk && accept_share));
  const at = String(row.accepted_at || "").trim() || null;
  return {
    accept_kvkk,
    accept_share,
    accepted,
    accepted_at: accepted ? at : null,
    required: row.required !== false,
    warning: row.warning || CONSENT_WARNING,
    kvkk_title: row.kvkk_title || KVKK_TITLE,
    kvkk_text: row.kvkk_text || KVKK_TEXT,
    share_title: row.share_title || SHARE_TITLE,
    share_text: row.share_text || SHARE_TEXT,
  };
}

export function locationConsentAccepted(raw) {
  return Boolean(normalizeLocationConsent(raw).accepted);
}

export function validateLocationConsent(req) {
  const row = req && typeof req === "object" ? req : {};
  const kvkk = Boolean(row.accept_kvkk || row.kvkk);
  const share = Boolean(row.accept_share || row.kk);
  if (!kvkk || !share) return "KVKK (K) ve konum paylaşımı (KK) sözleşmelerini işaretleyin.";
  return null;
}

export function locationConsentPayload() {
  return { accept_kvkk: true, accept_share: true };
}

export function normalizeLocationSignal(raw) {
  const row = raw && typeof raw === "object" ? raw : {};
  let tone = row.tone;
  let ok = row.ok;
  if (ok === true) tone = tone || "green";
  else if (ok === false) tone = tone || "red";
  else if (tone !== "green" && tone !== "red") {
    tone = "amber";
    ok = ok ?? null;
  }
  const label =
    row.label ||
    (tone === "green" ? "Konum alındı" : tone === "red" ? "Konum alınamadı" : "Konum bekleniyor");
  return { ok: ok === true ? true : ok === false ? false : null, at: row.at || null, tone, label };
}

export function locationUnavailablePayload(reason) {
  const text = String(reason || "").trim();
  return { reason: (text || "Konum izni kapalı veya GPS alınamadı.").slice(0, 200) };
}
